stdlib/internals.ncl

```diff
@@ -15,80 +15,71 @@
-      else
-          %blame% label,
+    if %typeof% value == `Function then
+      (fun x => %assume% codomain (%go_codom% label) (value (%assume% domain (%chng_pol% (%go_dom% label)) x)))
+    else
+      %blame% label,
 
   "$forall_var" = fun sealing_key label value =>
-      let current_polarity = %polarity% label in
-      let polarity = (%lookup_type_variable% sealing_key label).polarity in
-      if polarity == current_polarity then
-          %unseal% sealing_key value (%blame% label)
-      else
-          # Here, we know that this term should be sealed, but to give the right
-          # blame for the contract, we have to change the polarity to match the
-          # polarity of the `Forall`, because this is what's important for
-          # blaming polymorphic contracts.
-          %seal% sealing_key (%chng_pol% label) value,
+    let current_polarity = %polarity% label in
+    let polarity = (%lookup_type_variable% sealing_key label).polarity in
+    if polarity == current_polarity then
+      %unseal% sealing_key value (%blame% label)
+    else
+      # Here, we know that this term should be sealed, but to give the right
+      # blame for the contract, we have to change the polarity to match the
+      # polarity of the `Forall`, because this is what's important for
+      # blaming polymorphic contracts.
+      %seal% sealing_key (%chng_pol% label) value,
 
   "$forall" = fun sealing_key polarity contract label value =>
     contract (%insert_type_variable% sealing_key polarity label) value,
 
   "$enums" = fun case label value =>
-      if %typeof% value == `Enum then
-          %assume% case label value
-      else
-          %blame% (%label_with_message% "not an enum tag" label),
+    if %typeof% value == `Enum then
+      %assume% case label value
+    else
+      %blame% (%label_with_message% "not an enum tag" label),
 
   "$enum_fail" = fun label =>
-      %blame% (%label_with_message% "tag not included in the enum type" label),
+    %blame% (%label_with_message% "tag not included in the enum type" label),
 
   "$record" = fun field_contracts tail_contract label value =>
     if %typeof% value == `Record then
       # Returns the sub-record of `l` containing only those fields which are not
       # present in `r`. If `l` has a sealed polymorphic tail then it will be
       # preserved.
-      let field_diff = fun left right => array.fold_left
-        (fun acc field =>
-          if %has_field% field right then
-            acc
-          else
-            %record_insert% field acc (left."%{field}"))
-        (%record_empty_with_tail% left)
-        (%fields% left)
+      let field_diff =
+        fun left right =>
+          array.fold_left (fun acc field =>
+            if %has_field% field right then
+              acc
+            else
+              %record_insert% field acc (left."%{field}")) (%record_empty_with_tail% left) (%fields% left)
```

Comment:
Is there a character limit or any trigger for removing newlines like these and putting a mult-iline application like that back on one line?

Reply:
No, I'm afraid there's no character counting (by design, IIRC). It's either multi-line, or not.

In the case of a conditional expression, it would be possible to at least split this distinction over the true and false branches, independently. So, for example, if the true branch was originally written on a single-line, then it would stay that way.

Reply:
Sorry, I might have not been totally clear: I'm not talking about the if-then-felse (which is fine!), but about the whole enclosing application of `fold_left`:

I think the original code was:

```diff
-     let field_diff = fun left right => array.fold_left
-       (fun acc field =>
-        ...FUN DEF...)
-       (%record_empty_with_tail% left)
-       (%fields% left)
+     let field_diff =
+       fun left right =>
+        array.fold_left (
+           ...FUN DEF...
+        ) (%record_empty_with_tail% left) (%fields% left)
```

In particular, the initial application was split such that each argument was on its own line, and everything indented. In particular the application is thus arguably multi-line. But in the formatted version, it got back on one line, which is in my opinion a bit less readable for long application, but I also wonder if there is a limit to this behavior. If not, we're at risk of squashing multi-line applications on one very very long line.

Reply:
Right, I see what you mean :+1: In the grammar, these are applicative chains. At the moment, the only formatting rule for them is to insert a space between the operands. Like with infix operator chains, I think it would be possible to do as you say, based on the multi-line-ness of the top-most applicative node in the tree.

Issue: https://github.com/tweag/topiary/issues/398